devito/ir/support/basic.py

```diff
@@ -426,7 +426,25 @@ def distance(self, other):
             else:
                 v = i - j
                 if v.is_Number and v.is_finite:
-                    return Vector(S.ImaginaryUnit)
+                    # If i and j are numbers, we append an (I) distance
+                    if i.is_Number and j.is_Number:
+                        return Vector(S.ImaginaryUnit)
+                    # If both i and j are not numbers, there may be dimension-dependent
+                    # dependencies so we append the distance
+                    else:
+                        ret.append(v)
+
+                # We are writing over an entire dimension but reading from one point.
+                # If there are overlaps between the two then we would have a dependency
+                # This is a conservative estimation as there are cases (example below)
+                # where we may or may not have a dependency given that we don't write
+                # depending on domain size, which is not compilation-time known
+
+                # For example:
+                # Eq(u[0, y], 1)
+                # Eq(u[1, y+1], u[0, 1])
+                elif i.is_Number and not j.is_Number:
```

Comment:
instead of `not j.is_Number` you can also use `j.is_Symbol`, they should be equivalent in this context

also you must cover the symmetric case:
```
elif (i.is_Number and j.is_Symbol) or (i.is_Symbol and j.is_Number):
```

Reply:
(parametrize test?)

Reply:
> instead of `not j.is_Number` you can also use `j.is_Symbol`, they should be equivalent in this context
> 
> also you must cover the symmetric case:
> 
> ```
> elif (i.is_Number and j.is_Symbol) or (i.is_Symbol and j.is_Number):
> ```

For this George and I figured that we only need `i.is_Number and j.is_Symbol` case since we would only have a dependency if we are reading from one location and writing over an entire dimension. The other case would have been caught previously. 

Consider the equations
```
eq0 = Eq(u[0,y], 1)
eq1 = Eq(u[1,1], u[0, y+1])
```
The dependency between reading from `u[0, y+1]` and writing to `u[0,y]` is caught by
```
v = i - j
if v.is_Number and v.is_finite:
    # If i and j are numbers, we append an (I) distance
    if i.is_Number and j.is_Number:
        return Vector(S.ImaginaryUnit)
    # If both i and j are not numbers, 
    # there may be dimension-dependent dependencies 
    # so we append the distance
    else:
        ret.append(v)
``` 
The dependency between reading from `u[0,y+1]` and writing to `u[1,1]` is imaginary. Hence the algorithm doesn't append anything to the distance vector

Reply:
I'm not so particularly convinced we shouldn't care about the symmetrict case

what if you have

```
Eq(u[0, 1], 1)
Eq(u[x, y], u[0, y])
```

?

Also:
* you're still using "not j.is_Number", instead of "is_Symbol"
* how is your solution any stronger and/or more robust than just having
```
if i.is_Symbol or j.is_Symbol:
   ..
```